src/olm/account/mod.rs

```diff
@@ -682,16 +721,167 @@ mod libolm {
+                Some(key) => {
+                    let mut ret = true.encode(writer)?;
+                    ret += key.encode(writer)?;
+
+                    ret
+                }
+            };
+
+            Ok(ret)
+        }
+    }
+
+    #[derive(Encode, Decode, Zeroize, ZeroizeOnDrop)]
+    pub(super) struct Pickle {
+        version: u32,
+        private_curve25519_key: Box<[u8; 32]>,
+        private_ed25519_key: Box<[u8; 64]>,
+        one_time_keys: Vec<OneTimeKey>,
+        opt_fallback_key: OptFallbackKey,
+    }
+
+    pub(super) const PICKLE_VERSION: u32 = 0x80000000;
+
+    impl From<&Account> for Pickle {
+        fn from(account: &Account) -> Self {
+            let one_time_keys: Vec<_> = account
+                .one_time_keys
+                .secret_keys()
+                .iter()
+                .map(|(_key_id, secret_key)| OneTimeKey { private_key: secret_key.to_bytes() })
+                .collect();
+
+            let fallback_key =
+                account.fallback_keys.fallback_key.as_ref().and_then(|f| f.try_into().ok());
+
+            Self {
+                version: PICKLE_VERSION,
+                private_curve25519_key: account.diffie_hellman_key.secret_key().to_bytes(),
+                private_ed25519_key: account.signing_key.expanded_secret_key(),
+                one_time_keys,
+                opt_fallback_key: OptFallbackKey { fallback_key },
+            }
+        }
+    }
+
+    impl TryFrom<Pickle> for Account {
+        type Error = LibolmPickleError;
+
+        fn try_from(pickle: Pickle) -> Result<Self, Self::Error> {
+            let mut one_time_keys = OneTimeKeys::new();
+
+            for (num, key) in pickle.one_time_keys.iter().enumerate() {
+                let secret_key = Curve25519SecretKey::from_slice(&key.private_key);
+                let key_id = KeyId(num as u64);
+                one_time_keys.insert_secret_key(key_id, secret_key, true);
+            }
+            one_time_keys.next_key_id = pickle.one_time_keys.len() as u64;
+
+            let fallback_keys = FallbackKeys {
+                key_id: 1,
+                fallback_key: pickle.opt_fallback_key.fallback_key.as_ref().map(|otk| otk.into()),
+                previous_fallback_key: None,
+            };
+
+            Ok(Self {
+                signing_key: Ed25519Keypair::from_expanded_key(&pickle.private_ed25519_key)?,
+                diffie_hellman_key: Curve25519Keypair::from_secret_key(
+                    &pickle.private_curve25519_key,
+                ),
+                one_time_keys,
+                fallback_keys,
+            })
+        }
+    }
+
+    pub fn expand_pickle_key(key: &[u8; 32], identity_key: &str) -> Box<[u8; 32]> {
+        let kdf: Hkdf<Sha256> = Hkdf::new(Some(identity_key.as_bytes()), key);
+        let mut key = [0u8; 32];
+
+        kdf.expand(b"dehydrated-device-pickle-key", &mut key)
+            .expect("We should be able to expand the 32 byte pickle key");
```

Comment:
Ah, we do avoid the IV reuse. Why is the identity key passed as a string here? We should probably concatenate the identity key to the info string instead of putting it into the salt, a mistake I made in the SDK a well.

Though a random IV/nonce would have been even better, this would avoid the IV reuse even if we re-upload the same pickle with a fresh set of one-time keys or for whatever other reason we might want to re-encrypt the dehydrated device pickle.